### DISCUSSION TITLE
What to do about old Hacktoberfest PRs?

### DISCUSSION THREAD

**Discussion by tianyizheng02** [TheAlgorithms/Python#8912] [Team Posts]
@cclauss Re: this point from [your recent comment](https://github.com/TheAlgorithms/Python/pull/8902#issuecomment-1657090712):
>  I could really use your help as a maintainer to clean up our pull requests ahead of Octoberfest [sic].

Given that a large portion (if not a large majority) of open PRs are from past Hacktoberfests, handling those PRs is the most effective way to massively reduce our backlog. However, most of those PRs have some problem or another that prevent us from merging them as is (no type hints, improper formatting, failing doctests, no doctests, etc). My worry is that these problems will never be resolved and reviewing these PRs would be a waste of time. Given that these PRs were (presumably) made mainly to reach the Hacktoberfest PR quota, I doubt that most if any of the PRs' authors care enough to revise a one- or two-year-old PR of theirs from an event that has long since concluded—that is, if they see our reviews at all.

With this in mind, what do you (or anyone else) think is the best way to handle these old Hacktoberfest PRs? Should we fix them ourselves and merge what we can? Or should we just close these PRs since most of them are lost causes anyway? On one hand, I don't like the former because it might set an undesirable precedent of contributors expecting us maintainers to fix their code for them. On the other hand, I don't like the latter either because it rejects many legitimate algorithm contributions, and merging them makes our jobs easier if/when future contributors try to contribute the same algorithms.

**Comment by cclauss** [comment]
I think we need to close those PRs that have little value or would require more work than it is worth to merge.   We must always be courteous to our contributors -- "Thank you for your contribution but...", "Can you please try to rebase your pull request to our current `master` branch" -- but ___we need to close all open PRs before October 1st___.  If you see PRs that cannot be rehabilitated, please close them with a courteous message.

Based on past experience, I think that we should **autoclose all new pull requests that are opened in the week** between September 24th and October 1st so that we start Hacktoberfest with a clean slate.  Are there ways that we should improve our contribution process or continuous integration tests to keep our code quality as high as possible?  We have earned more than 161K GitHub Stars by being a welcoming learning community that maintains high-quality algorithms.  Let's keep up that great tradition.

@TheAlgorithms/python_maintainers

**Reply by tianyizheng02** [reply]
> "Thank you for your contribution but...", "Can you please try to rebase your pull request to our current master branch" -- but we need to close all open PRs before October 1st. If you see PRs that cannot be rehabilitated, please close them with a courteous message.

For PRs that _can_ be rehabilitated (i.e., needs few to no fixes and only really needs to be rebased), we could leave a message asking the author to rebase and later close the PRs if the author doesn't respond by, say, September 1st. Does that sound like a reasonable timeframe?

> Based on past experience, I think that we should autoclose all new pull requests that are opened in the week between September 24th and October 1st so that we start Hacktoberfest with a clean slate.

Is there anything we should to facilitate that when the time comes? Update the README with a "We're not currently accepting PRs" notice, update algorithms-keeper to automatically close all PRs during that week, anything like that?

> Are there ways that we should improve our contribution process or continuous integration tests to keep our code quality as high as possible?

I think we need to be more vigilant in looking out for the following:
  - Alternate implementations of existing algorithms: As pointed out in #8098, we have many algorithms in the repo with different implementations in multiple files (and often with different names). Having implementations spread out like this makes it harder to spot duplicates (see #8812) and harder to compare implementations against each other (benchmarking, etc). Moving forward, I think we should ask contributors to add new implementations of existing algorithms in an existing file instead of making a new file.
  - How-tos: As pointed out in #6216, many algorithms (particularly in `machine_learning/`) are just tutorials on how to use `sklearn` or some other library's existing implementation of an algorithm. We must be stricter about not letting these through since they're antithetical to the purpose of this repo, though they're typically really easy to spot.
  - Plagiarism: This is particularly a problem during Hacktoberfest when people are looking for easy PRs to meet their quota. Obviously we should immediately reject plagiarized PRs, but they can sometimes be hard to spot and often require us to spend a not-insignificant amount of time on research, so we need to be wary of them too.

Automating checking for any one of these as part of our CI would go a long way (assuming they _can_ be automated), but I don't really have any ideas at the moment on how to do that.

Also, is there a way to manually trigger PR checks? It'd make it a lot easier to ensure that really old PRs that passed checks in the past still pass current checks.

**Reply by dhruvmanila** [reply]
> update algorithms-keeper to automatically close all PRs during that week, anything like that?

Yes, the bot can be updated to do that.

> Alternate implementations of existing algorithms:

I'm not sure but we could use some kind of automation to detect this using machine learning \cc @expressgradient which could either be a standalone CI run or integrate it into the algorithms-keeper bot.

> Also, is there a way to manually trigger PR checks?

Yes, you can [manually run a workflow](https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow) using `workflow_dispatch` event trigger. This needs to be added in all of existing workflow files for which you would want a manual trigger.

**Reply by ExpressGradient** [reply]
> I'm not sure but we could use some kind of automation to detect this using machine learning \cc @ExpressGradient which could either be a standalone CI run or integrate it into the algorithms-keeper bot.

Yes we can use the LLMs to detect the duplication/alternate implementation of the algorithms. I'd say it's best if we integrate this feature in keeper bot only.

**Reply by dhruvmanila** [reply]
@ExpressGradient (or anyone else) Are you interested in implementing it? If not, it's totally fine.

The repository is at https://github.com/TheAlgorithms/algorithms-keeper

**Reply by rohan472000** [reply]
I'll try to implement it..but before that plz specify what exact functionality supposed to be added, duplicate algo identification and any other PR with specific tags??

**Reply by tianyizheng02** [reply]
@cclauss Are we good to start auto-closing all new PRs?

**Reply by cclauss** [reply]
Yes please.  What text will we put in the close messages?

**Reply by tianyizheng02** [reply]
Perhaps something like
> Thank you for your contribution, but unfortunately we're not accepting new PRs at the moment. We're currently trying to clear our backlog of existing PRs in preparation for Hacktoberfest. If you want to contribute, please wait until after October 1 to do so.

**Reply by cclauss** [reply]
This is great!  There are currently 80+ open pull requests.  Please close them before October 1st starts.

This ONLY applies to the **Python** repo.  Other repos remain open to new pull requests.

Once Hacktoberfest starts, let’s keep our standards high.  All pull requests require tests, type hints, good variable and function names, etc.  Our code quality has helped our repo to achieve 170k GitHub ⭐️ so it is vital that we retain the high quality that visitors expect.

If submissions do not meet our well documented standards then add a link to `CONTRIBUTING.md` and close them.  Let’s also make sure that contributions are algorithms as defined in CONTRIBUTING.md and not how-to-use examples (except in `web programming`).  Please make sure that all messages courteous with please and thank you, etc. because we want to continue to encourage quality contributions from a broad audience.

We must try to keep the number of open PRs low as we proceed thru the busy month ahead.

**Reply by tianyizheng02** [reply]
By the way, I just found out that this year Hacktoberfest won't be providing free t-shirts. I don't know about you all, but I feel that the free t-shirts were a large incentive for people to participate. Maybe this means the influx of Hacktoberfest PRs will be less extreme this year?

**Reply by rohan472000** [reply]
But still digital reward and plantation on contributor name , these 2 incentives are there.

**Reply by cclauss** [reply]
I am OK with not getting the contributions from folks who only wanted a tee-shirt.
They usually want to fix typos (we have codespell for that) or change code formatting (we have psf/black for that).

**Comment by cclauss** [comment]
# It is October 1st in [Kiribati](https://en.wikipedia.org/wiki/Kiribati) so I have **closed the 50 remaining open PRs** and Hacktoberfest 2023 can officially begin...
